compute/etc/neon_collector.yml

```diff
@@ -94,6 +94,68 @@ metrics:
   query: |
     select sum(pg_database_size(datname)) as total from pg_database;
 
+- metric_name: getpage_wait_seconds_count
+  type: counter
+  help: 'Number of getpage requests'
+  values: [getpage_wait_seconds_count]
+  query_ref: neon_perf_counters
+
+- metric_name: getpage_wait_seconds_sum
+  type: counter
+  help: 'Time spent in getpage requests'
+  values: [getpage_wait_seconds_sum]
+  query_ref: neon_perf_counters
+
+- metric_name: getpage_prefetch_requests_total
+  type: counter
+  help: 'Number of getpage issued for prefetching'
+  values: [getpage_prefetch_requests_total]
+  query_ref: neon_perf_counters
+
+- metric_name: getpage_sync_requests_total
+  type: counter
+  help: 'Number of synchronous getpage issued'
+  values: [getpage_sync_requests_total]
+  query_ref: neon_perf_counters
+
+- metric_name: getpage_prefetch_misses_total
+  type: counter
+  help: 'Number of prefetch misses (XXX what does this mean?)'
```

Comment:
What would it take to figure out what it means?

Reply:
Good question. Read the code and figure it out I guess.. So, the code where this counter is incremented is here:

https://github.com/neondatabase/neon/blob/7cfd116856d07ab34a63be6c221cc7f9ffd92a58/pgxn/neon/pagestore_smgr.c#L2817

I added this as a metric because we also display that in `EXPLAIN (PREFETCH, ANALYZE)` output:
```
neondb=> explain (analyze, prefetch) select count(*) from x;
                                                               QUERY PLAN                                                                
-----------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=77576.92..77576.93 rows=1 width=8) (actual time=5628.052..5648.708 rows=1 loops=1)
   Prefetch: hits=39008 misses=0 expired=0 duplicates=39008
   ->  Gather  (cost=77576.71..77576.92 rows=2 width=8) (actual time=5627.916..5648.702 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         Prefetch: hits=39008 misses=0 expired=0 duplicates=39008
         ->  Partial Aggregate  (cost=76576.71..76576.72 rows=1 width=8) (actual time=5619.183..5619.184 rows=1 loops=3)
               Prefetch: hits=39008 misses=0 expired=0 duplicates=39008
               ->  Parallel Seq Scan on x  (cost=0.00..69062.97 rows=3005497 width=0) (actual time=3.136..5490.095 rows=2404397 loops=3)
                     Prefetch: hits=39008 misses=0 expired=0 duplicates=39008
 Planning Time: 0.032 ms
 Execution Time: 5648.735 ms
(12 rows)
```
Exposing the same counters as metrics seemed like a good starting point.

Looking at the code in neon_read_at_lsnv(), it's not clear to me when 'misses' can ever be non-zero. AFAICS, even in the synchronous case, when we will block and wait for the result, we always create a "prefetch request", so we should never get a prefetch miss. I wonder if this used to work differently, but refactorings made as part of the v17 vectorized read work broke the accounting. @MMeent what do you think?

Reply:
> I updated the description in #9116: https://github.com/neondatabase/neon/pull/9116/files#diff-303a60096b4cd13a5e7ac9b130f373219fe37be66886c99dffcaae99f97b4027R60-R64

> > `Total number of readahead misses; consisting of either prefetches that don't satisfy the LSN bounds once the prefetch got read by the backend, or cases where somehow no readahead was issued for the read.`

Ah, thanks, I'll copy that to this PR.

Reply:
> IIUC, the accounting logic here hasn't changed from how it worked previously.

Hmm, the code was different before commit 78938d1b591b33d23495a0edb8b123cc5cac6a27. Before that, there was no `prefetch_register_bufferv` call in `neon_read_at_lsnv`, so a "synchronous" non-prefetched read would reach that `pgBufferUsage.prefetch.misses += 1;` line. After 78938d1b591b33d23495a0edb8b123cc5cac6a27, we always do `prefetch_register_bufferv` first, so you practically never reach that line.

So I believe we have a bug in how those counters are updated for EXPLAIN (PREFETCH), introduced in commit 78938d1b591b33d23495a0edb8b123cc5cac6a27. It only affects those numbers in EXPLAIN output, though, the logic is otherwise correct. I'll open a separate issue to track that.